pkg/hive/cell/health.go

```diff
@@ -0,0 +1,249 @@
+	All() []Status
+
+	// Get returns a copy of a modules status, by module ID.
+	// This includes unknown status for modules that have not reported a status yet.
+	Get(string) *Status
+
+	// forModule creates a moduleID scoped reporter handle.
+	forModule(string) StatusReporter
+}
+
+// Update is an event that denotes the change of a modules health state.
+type Update struct {
+	Level
+	ModuleID string
+	Message  string
+	Err      error
+}
+
+// Status is a modules last health state, including the last update.
+type Status struct {
+	Update
+	Stopped     bool
+	LastOK      time.Time
+	LastUpdated time.Time
+}
+
+func (s *Status) String() string {
+	var sinceLast string
+	if s.LastUpdated.IsZero() {
+		sinceLast = "never"
+	} else {
+		sinceLast = time.Since(s.LastUpdated).String() + " ago"
+	}
+	return fmt.Sprintf("Status{ModuleID: %s, Level: %s, Since: %s, Message: %s, Err: %v}",
+		s.ModuleID, s.Level, sinceLast, s.Message, s.Err)
+}
+
+// NewHealthProvider starts and returns a health status which collects emitted
+// health status declarations and updates module health state.
+func NewHealthProvider() *HealthProvider {
+	p := &HealthProvider{
+		updates:        make(chan Update, updatesBufferSize),
+		moduleStatuses: make(map[string]Status),
+	}
+	p.running.Store(true)
+
+	p.obs, p.emit, p.complete = stream.Multicast[Update]()
+
+	// Listen for updates, use buffered channel to avoid blocking.
+	go func() {
+		for u := range p.updates {
+			prev := func() Status {
+				p.mu.Lock()
+				defer p.mu.Unlock()
+				t := time.Now()
+				prev := p.moduleStatuses[u.ModuleID]
+				ns := Status{
+					Update:      u,
+					LastUpdated: t,
+				}
+				switch u.Level {
+				case StatusOK:
+					ns.LastOK = t
+				case StatusStopped:
+					// If Stopped, set that module was stopped and preserve last known status.
+					ns = prev
+					ns.Stopped = true
+				}
+				p.moduleStatuses[u.ModuleID] = ns
+				log.WithField("status", ns.String()).Debug("Processed new health status")
+				return prev
+			}()
+			p.processed.Add(1)
+			if prev.Stopped {
+				log.Warnf("module %q reported health status after being Stopped", u.ModuleID)
+			}
+		}
+	}()
+
+	return p
+}
+
+const providerDrainPollingInterval = 100 * time.Millisecond
+
+// Finish stops the status provider, and waits for all updates to be processed or
+// returns an error if the context is cancelled first.
+func (p *HealthProvider) Stop(ctx context.Context) error {
+	defer close(p.updates)
+	p.running.Store(false) // following this, no new reporters will send.
+	// Wait for any remaining reporters to finish.
+	p.wg.Wait()
+	// Wait for all updates to be processed, ensuring a consistent state.
+done:
+	for {
+		select {
+		case <-inctimer.After(providerDrainPollingInterval):
+			processed := p.processed.Load()
+			queued := p.queued.Load()
+			// Wait for existing reporters to finish.
+			if processed >= queued {
+				break done
```

Comment:
why the label? Could we just return at this point?

Reply:
this logic isnt necessary anymore following removing buffered channel.